### PR TITLE
[php-symfony] Fix return type in model setters.

### PR DESCRIPTION
**Pull request by reznikartem**
Previously return type was same, as method arguments. It`s wrong, and cause errors like
"Return value of Foo::setSuccess() must be of the type bool, object returned"
We cant use self and current {{classname}} as return type, because that can break class inheritance. So, better remove type hint on setters, until PHP-devs dont make realization for return static

- [+] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [+] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [+] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [+] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [+] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@jebentier (2017/07), @dkarlovi (2017/07), @mandrean (2017/08), @jfastnacht (2017/09), @ackintosh (2017/09) ❤️, @ybelenko (2018/07), @renepardon (2018/12)

### CODE REVIEW

**Comment by ybelenko** [comment]
> We cant use self and current {{classname}} as return type, because that can break class inheritance.

Hmm... Can you give an example with broken inheritance? I've seen many packages with chain like setters `$item->setValueA('foo')->setValueB('bar');` and they work as expected.

I'm not sure that you can break something with `return $this;` in php when returned type is class of current method.

**Comment by reznikartem** [comment]
> UPD: Or maybe you mean that you can't override class method in inherited class, because it should return the same type...

Yep, i'm about code like that

```php
<?php
class A
{
    public function foo(): self
    {
        return $this;
    }
}

class B extends A
{
    public function foo(): self
    {
        //do something
        return parent::foo();
    }
}

(new B())->foo();
```

It produces error `PHP Fatal error:  Declaration of B::foo(): B must be compatible with A::foo(): A`

We can write this code like that 

```php
<?php
class A
{
    public function foo(): A
    {
        return $this;
    }
}

class B extends A
{
    public function foo(): A
    {
        //do something
        return parent::foo();
    }
   
    public function bar()
    {
    }
}

(new B())->foo();
```

But it's wrong too, i think? IDE will not know about method `bar()`, that we able to call after `foo()` in class `B`

I don't know - is that can be problem in open api generator, but suggest to remove return type hint in setters for avoid possible problems with this code in future. 

Anyway - current version of setters in models don't work at all. We update codegenerator today, and get fatal errors :(

**Comment by ybelenko** [comment]
> But it's wrong too, i think? IDE will not know about method bar(), that we able to call after foo() in class B

Or we can set abstract class or interface as return type. In current case it's `Controller` class from symfony framework. It gives something, at least public methods from `Controller` class which is better than nothing.

**Comment by ybelenko** [comment]
Sorry, my bad, I confused it with Controllers while we're talking about models. They don't extend any class or interface. Let's stick with first implementation and return base type class. So for `Order` class it will be `setShipdate(): Order{}` and for inherited class will be the same.

**Comment by reznikartem** [comment]
In this PR i'm change model class, not controller. By default, model don't inherited from any class (for example - this Order class https://github.com/OpenAPITools/openapi-generator/blob/0b7c075d53d217714dc4092854bfacf60e1dc9c5/samples/server/petstore/php-symfony/SymfonyBundle-php/Model/Order.php )

But it can be inherited from another model, that can be inherited from another one model etc... Because of that, i'm suppose that PHPDoc annotation ‘@return $this‘ is don't bad variant?

**Comment by reznikartem** [comment]
Is it possible in this mustache file to get first class in chain of models? If, there will be class ‘C‘ that extends class ‘B‘, that extends class ‘A‘? We need to write same type hint ‘foo(): A‘ for all three models

**Comment by ybelenko** [comment]
> Is it possible in this mustache file to get first class in chain of models? If there will be class ’C’ that extends class ‘B‘, that extends class ‘A‘? We need to write same type hint ‘foo(): A‘ for all three models

As far as I know, team members never thought that somebody will override models. It's very unlikely that you need to extend models while you can change model template and generate whatever you want. 

API controllers is different because we cannot generate 100% complete implementation, so user needs to write his own code in extended class.

**Comment by reznikartem** [comment]
I'm don't need it too in openapi models. I'm just afraid, that it can be a problem for someone else in the future. If you think, that it will be better, i can add type hint  ": self" for setters in tomorrow commit for this PR.

**Comment by wing328** [comment]
> Is it possible in this mustache file to get first class in chain of models? If there will be class ’C’ that extends class ‘B‘, that extends class ‘A‘? We need to write same type hint ‘foo(): A‘ for all three models

Are you referring to the model inheritance? I think PHP client has better support on that. Not sure about PHP server generators.

**Comment by reznikartem** [comment]
Add return `self` type hint in last commit.

Btw... In composer.json, that generates by openapi generator says:

```json
 "require": {
        "php": ">=5.4",
}
```

but now, i think, there must be something like >= 7.1 (7.0 - for return type-hints, and 7.1 for nullable return type-hints)?

**Comment by ybelenko** [comment]
> but now, i think, there must be something like >= 7.1 (7.0 - for return type-hints, and 7.1 for nullable return type-hints)?

You're absolutely right. Issue #1246 explains that we need to migrate to PHP 7 in all PHP generators. It means update `composer.json` , `.travis.yml`, set type hints, update documentation etc.

**Comment by ybelenko** [comment]
I've checked models in current PHP client. Looks like:
https://github.com/OpenAPITools/openapi-generator/blob/3bbaedd9bc086028e7b1be973a30e04870962e5d/samples/client/petstore/php/OpenAPIClient-php/lib/Model/BigCat.php#L238-L259

Class `BigCat` extends `Cat`. It does return `$this`, but doesn't use return type hint maybe because of the problems we talked before.

**Comment by reznikartem** [comment]
@ybelenko done

**Comment by ybelenko** [comment]
CI fail isn't related to PR, obviously the same codebase already passed all tests in 0b7c075.

**Comment by ybelenko** [comment]
Now I'm facing that issue myself. I wanna extend generated model with internal database fields which public audience shouldn't be aware of.

Beside broken VSCode autocomplete (when I call parent method then child method in a chain) I cannot use interfaces with such methods. For instance, when I have interface with `setName(): self` and I implement it in parent class but declare it in child class I got error that child method doesn't match interface because it returns parent class instead of self.

Long story short. Don't know how to write chain setters with complex inheritance. Works great for standalone final class however.

**Comment by wing328** [comment]
https://github.com/OpenAPITools/openapi-generator/pull/11810 has been merged to support Symfony 6. Please give it a try and let us know if you've any feedback or question.
